internal/convert/registryv1.go

```diff
@@ -274,15 +267,18 @@ func Convert(in RegistryV1, installNamespace string, targetNamespaces []string)
 		permissions = nil
 	}
 
-	for _, permission := range permissions {
-		saName := saNameOrDefault(permission.ServiceAccountName)
-		name, err := generateName(fmt.Sprintf("%s-%s", in.CSV.Name, saName), permission)
-		if err != nil {
-			return nil, err
+	for _, ns := range targetNamespaces {
+		for _, permission := range permissions {
+			saName := saNameOrDefault(permission.ServiceAccountName)
+			name, err := generateName(fmt.Sprintf("%s-%s", in.CSV.Name, saName), permission)
+			if err != nil {
+				return nil, err
+			}
+			roles = append(roles, newRole(ns, name, permission.Rules))
+			roleBindings = append(roleBindings, newRoleBinding(ns, name, name, installNamespace, saName))
```

Comment:
What is expected to happen if the specified namespace set is all namespaces (i.e `""` or `"*"`)?

Reply:
`"*"` is not a valid namespace name, and (correct me if I'm wrong) I don't think there is any existing code in OLM or Kubernetes that treats `"*"` as all namespaces?

If `targetNamespaces == []string{""}`, that has typically meant "please watch all namespaces"

Separate but related question: what if `targetNamespaces == []string{"", "foo"}`? Perhaps we should have a CRD schema validation that forbids empty string in this list, and document that empty/null list is the way to get "all namespaces".

Reply:
> `"*"` is not a valid namespace name, and (correct me if I'm wrong) I don't think there is any existing code in OLM or Kubernetes that treats `"*"` as all namespaces?
> 

I don't think there is - I've seen "*" used before in RBAC to mean "all apis" within a group, but I don't recall ever seeing it used in the namespace context so I'm happy to concede that I'm wrong with that being a valid "all namespaces" notation.

> If `targetNamespaces == []string{""}`, that has typically meant "please watch all namespaces"
> 

So then in this case we should do what?
- List all namespaces and create `Role`+`RoleBindings`?
- Escalate all permissions to `ClusterRole` + `ClusterRoleBindings`?
- Something else?

My reason for calling this out is that I think if `""` is specified the role+rolebinding creation in this logic will fail.

> Separate but related question: what if `targetNamespaces == []string{"", "foo"}`? Perhaps we should have a CRD schema validation that forbids empty string in this list, and document that empty/null list is the way to get "all namespaces".

+1 to validating that `""` doesn't appear in a list of specific namespaces

Reply:
As Joe mentioned, we do have validation in place that checks if `len(targetNs) == 1` and if it is "", if so we set to AllNamespace mode and elevate the Roles and RoleBindings to ClusterRoles and ClusterRoleBindings respectively. 

https://github.com/operator-framework/rukpak/blob/a990f15b67ffc35453801e814979b91a8ebab715/internal/convert/registryv1.go#L265

> what if targetNamespaces == []string{"", "foo"}

In this case our validation would check if MultiNamespace mode is supported or not. If not, it would error - https://github.com/operator-framework/rukpak/blob/a990f15b67ffc35453801e814979b91a8ebab715/internal/convert/registryv1.go#L161. But a corner case is:
- Multinamespace mode is set to true
- And target ns has a list with "" in it
Will add a validation as discussed to ensure - if len(targetNs) > 1, "" shouldn't appear in the list. 

On that note, just to confirm can MultiNamespace be set to true, and the targetNs list just have a single value? 
Currently, the answer is no. Because in that case, we expect SingleNs to be set instead. I am not aware of the reasoning behind differentiating single and multiNs in this context, but just checking if that's how we are handling in v0.

Reply:
>As Joe mentioned, we do have validation in place that checks if len(targetNs) == 1 and if it is "", we set to AllNamespace mode and elevate the Roles and RoleBindings to ClusterRoles and ClusterRoleBindings respectively.

Ah, I see now - thanks for providing a link! Even though I expanded the code to look for this I guess I didn't expand it enough and missed it 🤦

Reply:
Just to clarify, I'm suggesting that we use OpenAPI schema validation in the CRD to outright prevent `""` in the list.

Reply:
> Just to clarify, I'm suggesting that we use OpenAPI schema validation in the CRD to outright prevent "" in the list.

Fixed.